File: android/src/main/kotlin/com/example/vcare_payment_module/VcarePaymentModulePlugin.kt
package com.example.vcare_payment_module

import android.app.Activity
import android.content.Intent
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class VcarePaymentModulePlugin: FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private val SETUP_REQUEST_CODE = 1001

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "vcare_payment_module")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initStripeGateway" -> result.success("Gateway Initialized")
            "startStripeSetupSheet" -> {
                val intent = Intent(activity, StripeSetupActivity::class.java).apply {
                    putExtra("publishableKey", call.argument<String>("publishableKey"))
                    putExtra("clientSecret", call.argument<String>("clientSecret"))
                    putExtra("customerId", call.argument<String>("customerId"))
                    putExtra("ephemeralKey", call.argument<String>("ephemeralKey"))
                    putExtra("secretKey", call.argument<String>("secretKey"))
                    putExtra("clientName", call.argument<String>("clientName"))
                }
                activity?.startActivityForResult(intent, SETUP_REQUEST_CODE)
                result.success("Setup Sheet Launched")
            }
            else -> result.notImplemented()
        }
    }

    private fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == SETUP_REQUEST_CODE) {
            when (resultCode) {
                Activity.RESULT_OK -> {
                    val pmId = data?.getStringExtra("paymentMethodId")
                    val pmDetails = data?.getStringExtra("paymentMethodDetails")
                    val args = HashMap<String, String?>()
                    args["paymentMethodId"] = pmId
                    args["paymentMethodDetails"] = pmDetails
                    channel.invokeMethod("paymentSuccess", args)
                }
                Activity.RESULT_CANCELED -> channel.invokeMethod("paymentCanceled", null)
                Activity.RESULT_FIRST_USER -> channel.invokeMethod("paymentFailed", null)
            }
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener { requestCode, resultCode, data ->
            handleActivityResult(requestCode, resultCode, data)
            true
        }
    }

    override fun onDetachedFromActivityForConfigChanges() { activity = null }
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivity() { activity = null }
    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {}
}

File: ios/Classes/VcarePaymentModulePlugin.swift
import Flutter
import UIKit

public class VcarePaymentModulePlugin: NSObject, FlutterPlugin {
    
    var methodChannel: FlutterMethodChannel?
    var rootViewController: UIViewController? {
        return UIApplication.shared.delegate?.window??.rootViewController
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "vcare_payment_module", binaryMessenger: registrar.messenger())
        let instance = VcarePaymentModulePlugin()
        instance.methodChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initStripeGateway":
            result("Gateway Initialized")

        case "startStripeSetupSheet":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
                return
            }
            startSetupSheet(with: args)
            result("Setup Sheet Launched")

        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startSetupSheet(with args: [String: Any]) {
        let vc = StripeSetupViewController()
        vc.modalPresentationStyle = .fullScreen
        
        vc.publishableKey = args["publishableKey"] as? String
        vc.clientSecret = args["clientSecret"] as? String
        vc.customerId = args["customerId"] as? String
        vc.ephemeralKey = args["ephemeralKey"] as? String
        vc.secretKey = args["secretKey"] as? String
        vc.clientName = args["clientName"] as? String
        vc.merchantId = args["applePayMerchantID"] as? String
        
        vc.onSuccess = { paymentMethodId, paymentMethodDetails in
            let result: [String: String] = [
                "paymentMethodId": paymentMethodId ?? "",
                "paymentMethodDetails": paymentMethodDetails ?? ""
            ]
            self.methodChannel?.invokeMethod("paymentSuccess", arguments: result)
        }
        
        vc.onCancel = {
            self.methodChannel?.invokeMethod("paymentCanceled", arguments: nil)
        }
        
        vc.onFailure = { errorMsg in
            self.methodChannel?.invokeMethod("paymentFailed", arguments: errorMsg)
        }
        
        rootViewController?.present(vc, animated: true, completion: nil)
    }
}
